File: hash_generator.rb
# hash_generator.rb — Ruby версия

require 'digest'
require 'optparse'
require 'base64'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby hash_generator.rb [options] [input]"
  opts.on("-a", "--algo ALGO", "Алгоритм (sha512, sha256, sha384, sha1, md5)") { |a| options[:algo] = a }
  opts.on("-f", "--format FORMAT", "Формат (hex, base64)") { |f| options[:format] = f }
  opts.on("-F", "--file FILE", "Файл для хеширования") { |f| options[:file] = f }
  opts.on("-o", "--output FILE", "Сохранить хеш в файл") { |o| options[:output] = o }
  opts.on("-c", "--compare HASH", "Сравнить с эталонным хешем") { |c| options[:compare] = c }
end.parse!

algo = options[:algo] || 'sha512'
format = options[:format] || 'hex'
file_path = options[:file]
output = options[:output]
compare = options[:compare]
input = ARGV[0]

puts "\e[36m🔐 Hash Generator (Ruby)\e[0m"
puts "Алгоритм: #{algo.upcase}"

def compute_hash(data, algo)
  case algo
  when 'sha512' then Digest::SHA512.hexdigest(data)
  when 'sha256' then Digest::SHA256.hexdigest(data)
  when 'sha384' then Digest::SHA384.hexdigest(data)
  when 'sha1'   then Digest::SHA1.hexdigest(data)
  when 'md5'    then Digest::MD5.hexdigest(data)
  else raise "Неподдерживаемый алгоритм"
  end
end

def hash_file(filename, algo)
  total = File.size(filename)
  processed = 0
  digest = case algo
  when 'sha512' then Digest::SHA512.new
  when 'sha256' then Digest::SHA256.new
  when 'sha384' then Digest::SHA384.new
  when 'sha1'   then Digest::SHA1.new
  when 'md5'    then Digest::MD5.new
  else raise "Неподдерживаемый алгоритм"
  end
  File.open(filename, 'rb') do |f|
    while chunk = f.read(8192)
      digest.update(chunk)
      processed += chunk.bytesize
      if total > 1024 * 1024
        percent = (processed.to_f / total) * 100
        STDERR.print "\r⏳ Прогресс: %.1f%%" % percent
      end
    end
  end
  STDERR.puts if total > 1024 * 1024
  digest.hexdigest
end

hex_digest = if file_path
  puts "📂 Хеширование файла: #{file_path}"
  hash_file(file_path, algo)
elsif input
  puts "📝 Входные данные: #{input}"
  compute_hash(input, algo)
else
  puts "📝 Чтение из STDIN (Ctrl+D для окончания)"
  data = STDIN.read
  if data.empty?
    puts "\e[33m⚠️ Пустой ввод.\e[0m"
    exit 1
  end
  compute_hash(data, algo)
end

result = if format == 'base64'
  Base64.strict_encode64([hex_digest].pack('H*'))
else
  hex_digest
end

puts "\e[32mХеш (#{format}):\e[0m"
puts result

if compare
  if result == compare
    puts "\e[32m✅ Хеши совпадают!\e[0m"
  else
    puts "\e[31m❌ Хеши не совпадают!\e[0m"
  end
end

if output
  File.write(output, result + "\n")
  puts "\e[32m💾 Сохранено в #{output}\e[0m"
end

File: hash_generator.rs
// hash_generator.rs — Rust версия

use sha2::{Sha512, Sha256, Sha384, Sha1, Digest};
use md5::Md5;
use std::env;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::Path;
use hex;
use base64;

fn compute_hash(data: &[u8], algo: &str) -> String {
    match algo {
        "sha512" => {
            let mut hasher = Sha512::new();
            hasher.update(data);
            hex::encode(hasher.finalize())
        }
        "sha256" => {
            let mut hasher = Sha256::new();
            hasher.update(data);
            hex::encode(hasher.finalize())
        }
        "sha384" => {
            let mut hasher = Sha384::new();
            hasher.update(data);
            hex::encode(hasher.finalize())
        }
        "sha1" => {
            let mut hasher = Sha1::new();
            hasher.update(data);
            hex::encode(hasher.finalize())
        }
        "md5" => {
            let mut hasher = Md5::new();
            hasher.update(data);
            hex::encode(hasher.finalize())
        }
        _ => panic!("Неподдерживаемый алгоритм"),
    }
}

fn hash_file(filename: &str, algo: &str) -> Result<String, io::Error> {
    let mut file = File::open(filename)?;
    let total = file.metadata()?.len();
    let mut buffer = [0; 8192];
    let mut processed = 0;
    match algo {
        "sha512" => {
            let mut hasher = Sha512::new();
            loop {
                let n = file.read(&mut buffer)?;
                if n == 0 { break; }
                hasher.update(&buffer[..n]);
                processed += n as u64;
                if total > 1024 * 1024 {
                    let percent = (processed as f64 / total as f64) * 100.0;
                    eprint!("\r⏳ Прогресс: {:.1}%", percent);
                }
            }
            if total > 1024 * 1024 { eprintln!(); }
            Ok(hex::encode(hasher.finalize()))
        }
        _ => {
            // аналогично для других алгоритмов, но для краткости используем compute_hash на всём файле
            // (для больших файлов может быть неэффективно, но для демонстрации сойдёт)
            let mut data = Vec::new();
            file.read_to_end(&mut data)?;
            Ok(compute_hash(&data, algo))
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let mut algo = "sha512".to_string();
    let mut format = "hex".to_string();
    let mut file_path = None;
    let mut output = None;
    let mut compare = None;
    let mut input = None;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--algo" | "-a" => {
                algo = args[i+1].clone();
                i += 2;
            }
            "--format" | "-f" => {
                format = args[i+1].clone();
                i += 2;
            }
            "--file" | "-F" => {
                file_path = Some(args[i+1].clone());
                i += 2;
            }
            "--output" | "-o" => {
                output = Some(args[i+1].clone());
                i += 2;
            }
            "--compare" | "-c" => {
                compare = Some(args[i+1].clone());
                i += 2;
            }
            _ => {
                if !args[i].starts_with("-") {
                    input = Some(args[i].clone());
                }
                i += 1;
            }
        }
    }

    println!("\x1b[36m🔐 Hash Generator (Rust)\x1b[0m");
    println!("Алгоритм: {}", algo.to_uppercase());

    let hex_digest = if let Some(ref fpath) = file_path {
        println!("📂 Хеширование файла: {}", fpath);
        hash_file(fpath, &algo)?
    } else if let Some(ref text) = input {
        println!("📝 Входные данные: {}", text);
        compute_hash(text.as_bytes(), &algo)
    } else {
        println!("📝 Чтение из STDIN (Ctrl+D для окончания)");
        let mut buffer = String::new();
        io::stdin().read_to_string(&mut buffer)?;
        if buffer.is_empty() {
            eprintln!("\x1b[33m⚠️ Пустой ввод.\x1b[0m");
            std::process::exit(1);
        }
        compute_hash(buffer.as_bytes(), &algo)
    };

    let result = if format == "base64" {
        let bytes = hex::decode(&hex_digest)?;
        base64::encode(&bytes)
    } else {
        hex_digest
    };

    println!("\x1b[32mХеш ({}):\x1b[0m", format);
    println!("{}", result);

    if let Some(ref expected) = compare {
        if result == *expected {
            println!("\x1b[32m✅ Хеши совпадают!\x1b[0m");
        } else {
            println!("\x1b[31m❌ Хеши не совпадают!\x1b[0m");
        }
    }

    if let Some(ref outfile) = output {
        let mut f = File::create(outfile)?;
        writeln!(f, "{}", result)?;
        println!("\x1b[32m💾 Сохранено в {}\x1b[0m", outfile);
    }

    Ok(())
}
